code/drasil-gool/lib/GOOL/Drasil/State.hs

```diff
@@ -548,7 +553,33 @@ getWordIndex = gets (^. (contentsIndices . _2))
 resetIndices :: MethodState -> MethodState
 resetIndices = set contentsIndices (0,0)
 
+useVarName :: String -> MethodState -> MethodState
+useVarName v = over (varNames . at prefix) (Just . max nextSuffix . fromMaybe 0)
+  where (prefix, nextSuffix) = over _2 (maybe 0 (+1)) $ splitVarName v
+
+genVarName :: [String] -> String -> MS String
+genVarName candidates backup = do
+  used <- gets (^. varNames)
+  let
+    isAvailable (n,c) = maybe True (maybe (const False) (>=) c) $ Map.lookup n used
+    choice = foldr const (splitVarName backup) $ filter isAvailable $ map splitVarName candidates
+  bumpVarName choice
+
+genLoopIndex :: MS String
+genLoopIndex = genVarName ["i", "j", "k"] "i"
+
 -- Helpers
 
 ifElemError :: (Eq a) => a -> [a] -> String -> [a]
-ifElemError e es err = if e `elem` es then error err else e : es
\ No newline at end of file
+ifElemError e es err = if e `elem` es then error err else e : es
+
+-- Split the longest numerical (0-9) suffix from the rest of the string
+splitVarName :: String -> (String, Maybe Int)
+splitVarName = over _2 readMaybe . over both reverse . swap . span isDigit . reverse
```

Comment:
To avoid the 2 `reverse` calls, I think you can use `span (not . isDigit)` to immediately get the string component. Afterwards you can use `Bifunctor`s `second` to convert the remaining `String` into a number.

Reply:
So the problem is that if I have `"ab12cb34"`, then `span (not . isDigit)` will give me `("ab", "12cb34")`, whereas I want `("ab12cb", "34")`.

Reply:
This is on me, I will document this better 😄

But basically `splitVarName` is used for both `useVarName` (to mark a user-chosen name as used) and `genVarName`. And `genVarName` generates a name based on a list of candidates chosen by the user (e.g. `i`, `j`, `k`). So `splitVarName` needs to handle any user-provided variable name.